src/node/internal/zlib.d.ts

```diff
@@ -2,6 +2,24 @@ import { owner_symbol, type Zlib } from 'node-internal:internal_zlib_base';
 
 export function crc32(data: ArrayBufferView, value: number): number;
 
+export type CompressCallback<ErrT, BufT> = (
```

Comment:
This needs to be (err, undefined) | (null, val)

Reply:
Why is `| undefined` better than `?` in this context?

Reply:
So, this type allows the following callback output: `(new Error(), Buffer.from('aa'))` but in reality, we can't have an error and a buffer at the same time. Therefore you need to fix the type to something like this

```
export type CompressCallback<T = ArrayBufferView> = ((err: Error, val: undefined) => void | ((err: null, val: T) => void)
```

Reply:
I don't think that type is right either. It would allow the callback signature to be `callback(err: undefined, val: Buffer)`, i.e. a callback that can only handle success.

Reply:
so, this is for our use case. this type is not exposed to outside, so null vs. undefined does not really make a difference for us.

Reply:
Sorry let's forget about null or undefined. 

My definition:
type X = (T | null, U | null) -> void

Type X wrongly allows (null, null) as arguments. That's bad. 

Your definition 
type Y = (T, nil) -> void | (nil, U) -> void 

Type Y allows me to implement this callback. 
f(a: nil, b: U) -> void 

This callback will never be able to handle errors properly. Yet it's permitted by type Y

Reply:
> This callback will never be able to handle errors properly. Yet it's permitted by type Y

I don't follow?

Reply:
https://www.typescriptlang.org/play/?#code/C4TwDgpgBAwg9gWzAJwgZzTAhgGxwIywGMBrKAXigAoqJlk5kAuKAUXsYBopU0BXHMBZ8AdgBMIAMwCWIiGICUFAHxQAbnGmKoAH2q0OzKKIky5Y7rwFCoaYMlkBzJeVUatCgNwAob5NFEwNJwIlCyGiQQVES4BMQkLPBIvJixhKRKAN7eUFAA9Hm2iND4cGIgUAAWdNDeAL6+-iKBwaESCHDYeOkkBgxGJlKy8pbo1ix2DiLOUNm5BVAAKpXSaFBNLSFha3CSUKCQsIgo6Knd8VD4fMBhNzEiAOQ3lVjiONB0-WtYNwCCeHAAO5oHJQIghNBwd4AOhwcEcVCsgi89V8C2WqygKHhyCwCCqWDWIjg+3A0FW-HQ3nCcEiVHanTS8S8QA

Reply:
Or on the flip side, if I try to work with that type, Typescript tells me both args resolve to "never":
https://www.typescriptlang.org/play/?#code/C4TwDgpgBAwg9gWzAJwgZzTAhgGxwIywGMBrKAXigAoqJlk5kAuKAUXsYBopU0BXHMBZ8AdgBMIAMwCWIiGICUFAHxQAbnGmKoAH2q0OzKKIky5Y7rwFCoaYMlkBzJeVUatCgNwAob5NFEwNJwIlCyGiQQVES4BMQkLPBIvJixhKRKAN7eUFAA9HlQACrIIE5QwHDGaNDAABbSaBXg0AhYkU2gkGhEDmDAFQ0iZBDEdVBYyI58CBAiA7xwOGroFVVyK8g5YZLUALJY9QB0yFjiiFRKqgAMRwCsWdu5MXjpJFRyAO5shlQARHBxvg4CA-gpuCYpLJ5F5tgBfKAQHA1KDZXLPNLxKiQszybh-ACSA0+jBIaAAhGCfLk4d5aX4AkEQlAJAg4NhXli6AwjDjoRYeOhrCw7A4RM5UdsCsUGk1-CJAsFQo0oHBdl1oEkUOhUpzSFB8HwBtIBjERAByAZ1M5iHDQbmMJqHCZ4bZEEJoJYQI44OCOKhWQSw+nhOCRKis9mYjKeIA

Reply:
Then I guess you need something like:

```
type CompressCallback = ((error: Error, result: undefined) => void) & ((error: undefined, result: string) => void);

function invoke(callback: CompressCallback) {
  // some body here 
}

function demoCallback(error: undefined, result: string | undefined) {
  // This function is of type CompressCallback but it can't handle errors at Allows
  console.log(result);
}

// This program has no type issues
invoke(demoCallback);
```

But I personally prefer something like this:


```
type CompressCallback = ((error: Error, result: undefined) => void) & ((error: undefined, result: string) => void);

function invoke(callback: CompressCallback) {
  // some body here 
}

function demoCallback(error: Error, result: undefined) : void;
function demoCallback(error: undefined, result: string) : void;
function demoCallback(error: Error | undefined, result: string | undefined) {
  // This function is of type CompressCallback but it can't handle errors at Allows
  console.log(result);
}

// This program has no type issues
invoke(demoCallback);
```

Reply:
Here's the fun bit... using the `&` type allows both an error and a result, just like my original type:
https://www.typescriptlang.org/play/?#code/C4TwDgpgBAwg9gWzAJwgZzTAhgGxwIywGMBrKAXigAoqJlk5kAuKAUXsYBopU0BXHMBZ8AdgBMIAMwCWIiGICUFAHxQAbnGmKoAMmq0OzKKIky5Y7rwFCoaYMlkBzJeVUatCgNwAob5NFEwNJwIlCyGiQQVES4BMQkLPBIvJixhKRKAN7eUFAA9Hm2iND4cGIgUAAWdNDeAL6+-iKBwaESCHDYeOkkBgxG7P1QAD7G4lKy8pbo1ix2DiKOI2Omk9rZuQVQACqV0mhQTS0hYQdwklCgkLCIKOip3fFQ+HzAYW8xIgDkb5VY4jhoHR+gcsG8AIJ4OAAdzQOSgRBCaDggIAdDg4I4qFZBF56r4trt9lAUJjkFgEFUsAcRHBLuBoPt+OhvOE4JEqO1Oml4l4gA